src/lib/ndpi_main.c

```diff
@@ -9758,18 +9758,26 @@ char *ndpi_strnstr(const char *haystack, const char *needle, size_t len)
  * Same as ndpi_strnstr but case-insensitive.
  * Please note that this function is *NOT* equivalent to strncasecmp().
  */
-const char * ndpi_strncasestr(const char *str1, const char *str2, size_t len) {
-  size_t str1_len = strnlen(str1, len);
-  size_t str2_len = strlen(str2);
-  int i; /* signed! */
+const char * ndpi_strncasestr(const char *s, const char *find, size_t len) {
 
-  for(i = 0; i < (int)(str1_len - str2_len + 1); i++){
-    if(str1[0] == '\0')
-      return NULL;
-    else if(strncasecmp(str1, str2, str2_len) == 0)
-      return(str1);
+  if (!s || !find) {
+    return NULL;
+  }
 
-    str1++;
+  const size_t find_len = strlen(find);
+
+  if (find_len == 0) {
+    return s;
+  }
+
+  const size_t s_len = strnlen(s, len);
+
+  const char *const end_of_search = s + s_len - find_len + 1;
+
+  for (; s < end_of_search; ++s) {
+    if (strncasecmp(s, find, find_len) == 0) {
+      return s;
+    }
```

Comment:
I guess it would be nice to add such check to get rid of unnecessary `strncasecmp` calls.

```
  for (; s < end_of_search; ++s) {
    if (tolower((u_int8_t)*s) == tolower((u_int8_t)find[0])) {
      if (strncasecmp(s, find, find_len) == 0) {
        return s;
      }
    }
  }
```

Reply:
I think that would improve performance in most cases.
I'll implement that.

Reply:
But I'm not sure whether I should cast `char` to `unsigned char` for `tolower` or not.
Do you have any arguments in favor of casting?

Reply:
no cast, the prototype is `int tolower(int c);`

Reply:
> no cast, the prototype is `int tolower(int c);`

Oh, shit, I'm sorry. For some reason I thought the argument type was `unsigned char`.

Reply:
Oh. I found the explanation in `man tolower`:

> If c is neither an unsigned char value nor EOF, the behavior of these functions is undefined.

So we should cast.

Reply:
However, since we're using `char` in the code, I think it's better to cast to `unsigned char` rather than `u_int8_t`.

Reply:
> Oh. I found the explanation in `man tolower`:
> 
> > If c is neither an unsigned char value nor EOF, the behavior of these functions is undefined.
> 
> So we should cast.

It is possible that the parameter is EOF? If not, we can cast without any issues; otherwise we should at least check for it, IMO

Reply:
> It is possible that the parameter is EOF? If not, we can cast without any issues; otherwise we should at least check for it, IMO

I think that placing EOF inside a string is so wrong that nobody would do that.

Reply:
I've added the optimization.
Since `find_len` is guaranteed to be greater than zero, we can call `strncasecmp` on the remaining parts of the strings without the first characters.
It should improve performance as well.